storage/replica.go

```diff
@@ -1681,42 +1686,65 @@ func defaultProposeRaftCommandLocked(r *Replica, p *pendingCmd) error {
 				return err
 			}
 
-			return r.withRaftGroupLocked(true, func(raftGroup *raft.RawNode) error {
-				return raftGroup.ProposeConfChange(raftpb.ConfChange{
-					Type:    changeTypeInternalToRaft[crt.ChangeType],
-					NodeID:  uint64(crt.Replica.ReplicaID),
-					Context: encodedCtx,
-				})
+			return r.withRaftGroupLocked(true, func(raftGroup *raft.RawNode) (bool, error) {
+				// We're proposing a command here so there is no need to wake the
+				// leader if we were quiesced.
+				r.unquiesceLocked()
+				return false, /* !unquiesce */
+					raftGroup.ProposeConfChange(raftpb.ConfChange{
+						Type:    changeTypeInternalToRaft[crt.ChangeType],
+						NodeID:  uint64(crt.Replica.ReplicaID),
+						Context: encodedCtx,
+					})
 			})
 		}
 
 		hasSplit = ict.GetSplitTrigger() != nil
 	}
 
-	return r.withRaftGroupLocked(true, func(raftGroup *raft.RawNode) error {
+	return r.withRaftGroupLocked(true, func(raftGroup *raft.RawNode) (bool, error) {
 		if log.V(4) {
 			log.Infof(r.ctx, "proposing command %x", p.idKey)
 		}
-		return raftGroup.Propose(encodeRaftCommand(string(p.idKey), data, hasSplit))
+		// We're proposing a command so there is no need to wake the leader if we
+		// were quiesced.
+		r.unquiesceLocked()
+		return false, /* !unquiesce */
+			raftGroup.Propose(encodeRaftCommand(string(p.idKey), data, hasSplit))
 	})
 }
 
-func (r *Replica) setQuiescent(v bool) {
+func (r *Replica) quiesce() {
 	r.mu.Lock()
-	r.setQuiescentLocked(v)
+	r.quiesceLocked()
 	r.mu.Unlock()
 }
 
-func (r *Replica) setQuiescentLocked(v bool) {
-	if r.mu.quiescent != v {
-		r.mu.quiescent = v
+func (r *Replica) quiesceLocked() {
+	if !r.mu.quiescent {
 		if log.V(3) {
-			if r.mu.quiescent {
-				log.Infof(r.ctx, "quiescing")
-			} else {
-				log.Infof(r.ctx, "unquiescing")
-			}
+			log.Infof(r.ctx, "quiescing")
+		}
+		r.mu.quiescent = true
+	}
+}
+
+func (r *Replica) unquiesceLocked() {
+	if r.mu.quiescent {
+		if log.V(3) {
+			log.Infof(r.ctx, "unquiescing")
 		}
+		r.mu.quiescent = false
+	}
+}
+
+func (r *Replica) unquiesceAndWakeLeaderLocked() {
+	if r.mu.quiescent {
+		if log.V(3) {
+			log.Infof(r.ctx, "unquiescing: waking leader")
+		}
+		r.mu.quiescent = false
+		_ = r.mu.internalRaftGroup.Propose(nil)
```

Comment:
This should be harmless because raft does generate these empty messages on its own, but it's inefficient because we watch for these messages (which are normally infrequent) and use them to trigger reproposals and other relatively expensive operations. We don't want to make these more common than they already are. We should send a different message to the leader to wake it up (or possibly a MsgProp containing no Entries, instead of an MsgProp containing one empty Entry).